contracts/governance/StkTruToken.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.6.10;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+import {VoteToken} from "./VoteToken.sol";
+import {ClaimableContract} from "../trusttoken/common/ClaimableContract.sol";
+import {ITruPriceOracle} from "./interface/ITruPriceOracle.sol";
+
+contract StkTruToken is VoteToken, ClaimableContract, ReentrancyGuard {
+    using SafeMath for uint256;
+
+    uint256 public constant MIN_STAKE_TIME = 14 days;
+
+    IERC20 public trustToken;
+    IERC20 public tusd;
+    ITruPriceOracle public oracle;
+    mapping(address => uint256) public unlockTime;
+
+    event OracleChanged(ITruPriceOracle newOracle);
+    event Stake(address indexed staker, uint256 amount, uint256 minted);
+    event Unstake(address indexed staker, uint256 burntAmount, uint256 truAmount, uint256 tusdAmount);
+
+    function initialize(
+        IERC20 _trustToken,
+        IERC20 _tusd,
+        ITruPriceOracle _oracle
+    ) public {
+        require(!initalized, "StkTruToken: Already initialized");
+        trustToken = _trustToken;
+        tusd = _tusd;
+        oracle = _oracle;
+        owner_ = msg.sender;
+        initalized = true;
+    }
+
+    function setOracle(ITruPriceOracle _oracle) external onlyOwner {
+        oracle = _oracle;
+        emit OracleChanged(_oracle);
+    }
+
+    function stake(uint256 amount) external {
+        unlockTime[msg.sender] = block.timestamp + MIN_STAKE_TIME;
+
+        uint256 amountToMint = totalSupply == 0 ? amount : totalSupply.mul(amount).div(value());
+        _mint(msg.sender, amountToMint);
+
+        require(trustToken.transferFrom(msg.sender, address(this), amount));
+
+        emit Stake(msg.sender, amount, amountToMint);
+    }
+
+    function unstake(uint256 amount) external nonReentrant {
+        require(balanceOf[msg.sender] >= amount, "StkTruToken: Insufficient balance");
+        require(unlockTime[msg.sender] <= block.timestamp, "StkTruToken: Stake is locked");
+
+        uint256 truAmount = trustToken.balanceOf(address(this)).mul(amount).div(totalSupply);
```

Comment:
since we will have TRU rewards and TRU in the staking pool, we need to calculate & transfer those separately here rather than use trustToken.balanceOf.

Reply:
But if TRU rewards are in staking pool as well, there shouldn't be any problem

Reply:
the reasoning is, un-staking is different than claiming. Stakers can incentive claim TRU at any time, but need to activate cooldown to un-stake staked TRU